mmdeploy/codebase/mmyolo/models/dense_heads/yolov5_head.py

```diff
@@ -0,0 +1,93 @@
+import copy
+from typing import List, Optional
+
+import torch
+from mmengine.config import ConfigDict
+from mmengine.structures import InstanceData
+from torch import Tensor
+
+from mmdeploy.codebase.mmdet import get_post_processing_params
+from mmdeploy.codebase.mmdet.models.layers import multiclass_nms
+from mmdeploy.core import FUNCTION_REWRITER
+
+
+@FUNCTION_REWRITER.register_rewriter(
+    func_name='mmyolo.models.dense_heads.yolov5_head.'
+    'YOLOv5Head.predict_by_feat')
+def yolov5_head__predict_by_feat(ctx,
+                                 self,
+                                 cls_scores: List[Tensor],
+                                 bbox_preds: List[Tensor],
+                                 objectnesses: Optional[List[Tensor]],
+                                 batch_img_metas: Optional[List[dict]] = None,
+                                 cfg: Optional[ConfigDict] = None,
+                                 rescale: bool = False,
+                                 with_nms: bool = True) -> List[InstanceData]:
+    assert len(cls_scores) == len(bbox_preds)
+    if objectnesses is None:
+        with_objectnesses = False
+    else:
+        with_objectnesses = True
+        assert len(cls_scores) == len(objectnesses)
+
+    cfg = self.test_cfg if cfg is None else cfg
+    cfg = copy.deepcopy(cfg)
+
+    num_imgs = len(batch_img_metas)
+    featmap_sizes = [cls_score.shape[2:] for cls_score in cls_scores]
+
+    mlvl_priors = self.prior_generator.grid_priors(
+        featmap_sizes, dtype=cls_scores[0].dtype, device=cls_scores[0].device)
+    flatten_priors = torch.cat(mlvl_priors)
+
+    mlvl_strides = [
+        flatten_priors.new_full(
+            (featmap_size.numel() * self.num_base_priors, ), stride)
+        for featmap_size, stride in zip(featmap_sizes, self.featmap_strides)
+    ]
+    flatten_stride = torch.cat(mlvl_strides)
+
+    # flatten cls_scores, bbox_preds and objectness
+    flatten_cls_scores = [
+        cls_score.permute(0, 2, 3, 1).reshape(num_imgs, -1, self.num_classes)
+        for cls_score in cls_scores
+    ]
+    flatten_bbox_preds = [
+        bbox_pred.permute(0, 2, 3, 1).reshape(num_imgs, -1, 4)
+        for bbox_pred in bbox_preds
+    ]
+
+    if with_objectnesses:
+        flatten_objectness = [
+            objectness.permute(0, 2, 3, 1).reshape(num_imgs, -1)
+            for objectness in objectnesses
+        ]
+    else:
+        flatten_objectness = [None for _ in range(len(featmap_sizes))]
+
+    cls_scores = torch.cat(flatten_cls_scores, dim=1).sigmoid()
+    flatten_bbox_preds = torch.cat(flatten_bbox_preds, dim=1)
+    flatten_objectness = torch.cat(flatten_objectness, dim=1).sigmoid()
+    bboxes = self.bbox_coder.decode(flatten_priors[None], flatten_bbox_preds,
+                                    flatten_stride)
+
+    # directly multiply score factor and feed to nms
+    scores = cls_scores * (flatten_objectness.unsqueeze(-1))
+    max_scores, _ = torch.max(scores, 1)
```

Comment:
We do not need to do the threshold here. There is a score threshold inside NMS.

Reply:
Got it. I will delete it